### PR TITLE
MAXIFS, MINIFS, AVERAGEIFS - Implementation and Tests

### PR DESCRIPTION
**Pull request by jniewerth**

### CODE REVIEW

**Comment by centic9** [comment]
Thanks for the contribution!

When I try to run tests on latest trunk with these changes I get three unit-test failures because of an NPE in Baseifs.java, can you take a look?

`java.lang.NullPointerException
at org.apache.poi.ss.formula.functions.Baseifs.accumulate(Baseifs.java:174)
at org.apache.poi.ss.formula.functions.Baseifs.aggregateMatchingCells(Baseifs.java:157)
at org.apache.poi.ss.formula.functions.Baseifs.evaluate(Baseifs.java:80)
at org.apache.poi.ss.formula.functions.Countifs.evaluate(Countifs.java:31)
at org.apache.poi.ss.formula.UserDefinedFunction.evaluate(UserDefinedFunction.java:59)
at org.apache.poi.ss.formula.OperationEvaluatorFactory.evaluate(OperationEvaluatorFactory.java:140)
at org.apache.poi.ss.formula.WorkbookEvaluator.evaluateFormula(WorkbookEvaluator.java:529)
at org.apache.poi.ss.formula.WorkbookEvaluator.evaluateAny(WorkbookEvaluator.java:258)
at org.apache.poi.ss.formula.WorkbookEvaluator.evaluate(WorkbookEvaluator.java:199)
at org.apache.poi.xssf.usermodel.BaseXSSFFormulaEvaluator.evaluateFormulaCellValue(BaseXSSFFormulaEvaluator.java:57)
at org.apache.poi.ss.formula.BaseFormulaEvaluator.evaluate(BaseFormulaEvaluator.java:109)
at org.apache.poi.ss.tests.formula.functions.TestCountifs.testBug56822(TestCountifs.java:118)`

`java.lang.NullPointerException
at org.apache.poi.ss.formula.functions.Baseifs.accumulate(Baseifs.java:174)
at org.apache.poi.ss.formula.functions.Baseifs.aggregateMatchingCells(Baseifs.java:157)
at org.apache.poi.ss.formula.functions.Baseifs.evaluate(Baseifs.java:80)
at org.apache.poi.ss.formula.functions.Countifs.evaluate(Countifs.java:31)
at org.apache.poi.ss.formula.UserDefinedFunction.evaluate(UserDefinedFunction.java:59)
at org.apache.poi.ss.formula.OperationEvaluatorFactory.evaluate(OperationEvaluatorFactory.java:140)
at org.apache.poi.ss.formula.WorkbookEvaluator.evaluateFormula(WorkbookEvaluator.java:529)
at org.apache.poi.ss.formula.WorkbookEvaluator.evaluateAny(WorkbookEvaluator.java:258)
at org.apache.poi.ss.formula.WorkbookEvaluator.evaluate(WorkbookEvaluator.java:199)
at org.apache.poi.hssf.usermodel.HSSFFormulaEvaluator.evaluateFormulaCellValue(HSSFFormulaEvaluator.java:201)
at org.apache.poi.ss.formula.BaseFormulaEvaluator.evaluate(BaseFormulaEvaluator.java:109)
at org.apache.poi.ss.tests.formula.functions.TestCountifs.testCallFunction(TestCountifs.java:83)`

`java.lang.NullPointerException
at org.apache.poi.ss.formula.functions.Baseifs.accumulate(Baseifs.java:174)
at org.apache.poi.ss.formula.functions.Baseifs.aggregateMatchingCells(Baseifs.java:157)
at org.apache.poi.ss.formula.functions.Baseifs.evaluate(Baseifs.java:80)
at org.apache.poi.ss.formula.functions.Countifs.evaluate(Countifs.java:31)
at org.apache.poi.ss.formula.UserDefinedFunction.evaluate(UserDefinedFunction.java:59)
at org.apache.poi.ss.formula.OperationEvaluatorFactory.evaluate(OperationEvaluatorFactory.java:140)
at org.apache.poi.ss.formula.WorkbookEvaluator.evaluateFormula(WorkbookEvaluator.java:529)
at org.apache.poi.ss.formula.WorkbookEvaluator.evaluateAny(WorkbookEvaluator.java:258)
at org.apache.poi.ss.formula.WorkbookEvaluator.evaluateReference(WorkbookEvaluator.java:785)
at org.apache.poi.ss.formula.SheetRefEvaluator.getEvalForCell(SheetRefEvaluator.java:48)
at org.apache.poi.ss.formula.SheetRangeEvaluator.getEvalForCell(SheetRangeEvaluator.java:74)
at org.apache.poi.ss.formula.LazyRefEval.getInnerValueEval(LazyRefEval.java:39)
at org.apache.poi.ss.formula.eval.OperandResolver.chooseSingleElementFromRef(OperandResolver.java:216)
at org.apache.poi.ss.formula.eval.OperandResolver.getSingleValue(OperandResolver.java:64)
at org.apache.poi.ss.formula.eval.RelationalOperationEval.evaluate(RelationalOperationEval.java:64)
at org.apache.poi.ss.formula.functions.Fixed2ArgFunction.evaluate(Fixed2ArgFunction.java:31)
at org.apache.poi.ss.formula.OperationEvaluatorFactory.evaluate(OperationEvaluatorFactory.java:138)
at org.apache.poi.ss.formula.WorkbookEvaluator.evaluateFormula(WorkbookEvaluator.java:529)
at org.apache.poi.ss.formula.WorkbookEvaluator.evaluateAny(WorkbookEvaluator.java:258)
at org.apache.poi.ss.formula.WorkbookEvaluator.evaluateReference(WorkbookEvaluator.java:785)
at org.apache.poi.ss.formula.SheetRefEvaluator.getEvalForCell(SheetRefEvaluator.java:48)
at org.apache.poi.ss.formula.SheetRangeEvaluator.getEvalForCell(SheetRangeEvaluator.java:74)
at org.apache.poi.ss.formula.LazyRefEval.getInnerValueEval(LazyRefEval.java:39)
at org.apache.poi.ss.formula.EvaluationConditionalFormatRule.unwrapEval(EvaluationConditionalFormatRule.java:391)
at org.apache.poi.ss.formula.EvaluationConditionalFormatRule.checkFormula(EvaluationConditionalFormatRule.java:401)
at org.apache.poi.ss.formula.EvaluationConditionalFormatRule.matches(EvaluationConditionalFormatRule.java:330)
at org.apache.poi.ss.formula.ConditionalFormattingEvaluator.getConditionalFormattingForCell(ConditionalFormattingEvaluator.java:182)
at org.apache.poi.ss.tests.usermodel.TestConditionalFormattingEval.getRulesFor(TestConditionalFormattingEval.java:190)
at org.apache.poi.ss.tests.usermodel.TestConditionalFormattingEval.testFormattingEvaluation(TestConditionalFormattingEval.java:120)`

File: poi/src/main/java/org/apache/poi/ss/formula/functions/Averageifs.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+     * need at least 3 arguments and need to have an odd number of arguments (average_range plus x*(criteria_range, criteria))
+     */
+    @Override
+    protected boolean hasInitialRange() {
+        return true;
+    }
+
+	@Override
+	protected Aggregator createAggregator() {
+		return new Aggregator() {
```
Comment (pjfanning) [inline poi/src/main/java/org/apache/poi/ss/formula/functions/Averageifs.java]: can you use remove the tabs? - most open source projects require spaces for indents - 4 spaces is the POI norm

File: poi/src/main/java/org/apache/poi/ss/formula/functions/Baseifs.java [hunk 2]
```diff
@@ -40,7 +40,17 @@
      * @return true if there should be a range argument before the criteria pairs
      */
     protected abstract boolean hasInitialRange();
-        
+  
+    /**
+     * Implements the details of a specific aggregation functions
+     */
+    protected static interface Aggregator {
+    	void addValue(ValueEval d);
```
Comment (pjfanning) [inline poi/src/main/java/org/apache/poi/ss/formula/functions/Baseifs.java]: likewise - the POI standard is 4 space indents

Comment (jniewerth) [reply poi/src/main/java/org/apache/poi/ss/formula/functions/Baseifs.java]: Thank you very much for reviewing and the comments. I fixed the whitespace and made the tests pass. Should I issue a new PR?

Comment (pjfanning) [reply poi/src/main/java/org/apache/poi/ss/formula/functions/Baseifs.java]: just commit the extra changes to your trunk branch and this PR will be automatically updated

**Comment by pjfanning** [comment]
thanks - merged using https://github.com/apache/poi/commit/47d80f116dd8ce24b9897355612d47bff62395d3

**Pull request closed without merging**